### PR TITLE
additional PR to fix unsolved issue

### PR DESCRIPTION
**Pull request by yingryic**
corresponds to #2781 @karlnapf

### CODE REVIEW

File: tests/unit/lib/SGSparseMatrix_unittest.cc [hunk 1]
```diff
@@ -27,18 +27,16 @@ using namespace Eigen;
 
 //function can generate both sparse and dense matrix accroding to sparse convention
 template<class MatrixType>
-MatrixType GenerateMatrix(float64_t sparseLevel, int32_t m, int32_t n, int32_t randSeed)
+void GenerateMatrix(float64_t sparseLevel, int32_t m, int32_t n, int32_t randSeed, MatrixType* matrix)
```
Comment (yingryic) [inline tests/unit/lib/SGSparseMatrix_unittest.cc]: SG dense matrix has no initialization upon construction, thus the template function cannot generate both sparse and dense matrix with zeros initialized. Thus I choose to initialize the matrix outside the function, and pass in it as an argument to set the nonzero entries.

**Comment by yingryic** [comment]
what does this mean @coveralls

File: src/shogun/lib/SGSparseMatrix.h [hunk 2]
```diff
@@ -126,11 +126,11 @@ template <class T> class SGSparseMatrix : public SGReferencedData
...
 		{
-			REQUIRE(i_row>=0, "index %d negative!\n", i_row);
-			REQUIRE(i_col>=0, "index %d negative!\n", i_col);
-			REQUIRE(i_row<num_features, "index should be less than %d, %d provided!\n",
+			REQUIRE(i_row>=0, "Index %d negative!\n", i_row);
+			REQUIRE(i_col>=0, "Index %d negative!\n", i_col);
+			REQUIRE(i_row<num_features, "Provided index (%d) is larger than number of rows (%d)\n",
 				num_features, i_row);
-			REQUIRE(i_col<num_vectors, "index should be less than %d, %d provided!\n",
+			REQUIRE(i_col<num_vectors, "Provided index (%d) is larger than number of columns (%d)\n",
```
Comment (karlnapf) [inline src/shogun/lib/SGSparseMatrix.h]: wrong order of numbers (num_vectors, i_col)

File: src/shogun/lib/SGSparseMatrix.h [hunk 3]
```diff
@@ -126,11 +126,11 @@ template <class T> class SGSparseMatrix : public SGReferencedData
 		 */
 		inline T& operator()(index_t i_row, index_t i_col)
 		{
-			REQUIRE(i_row>=0, "index %d negative!\n", i_row);
-			REQUIRE(i_col>=0, "index %d negative!\n", i_col);
-			REQUIRE(i_row<num_features, "index should be less than %d, %d provided!\n",
+			REQUIRE(i_row>=0, "Index %d negative!\n", i_row);
+			REQUIRE(i_col>=0, "Index %d negative!\n", i_col);
+			REQUIRE(i_row<num_features, "Provided index (%d) is larger than number of rows (%d)\n",
```
Comment (karlnapf) [inline src/shogun/lib/SGSparseMatrix.h]: wrong order

**Comment by karlnapf** [comment]
coveralls is the code coverage analysis

**Comment by karlnapf** [comment]
thanks for the patch! :)

Ill merge when you fixed the error message

**Comment by yingryic** [comment]
fixed, take a look @karlnapf

File: src/shogun/lib/SGSparseMatrix.h [hunk 4]
```diff
@@ -105,12 +105,12 @@ template <class T> class SGSparseMatrix : public SGReferencedData
 		 */
 		inline const T operator()(index_t i_row, index_t i_col) const
 		{
-			REQUIRE(i_row>=0, "index %d negative!\n", i_row);
-			REQUIRE(i_col>=0, "index %d negative!\n", i_col);
-			REQUIRE(i_row<num_features, "index should be less than %d, %d provided!\n",
-				num_features, i_row);
-			REQUIRE(i_col<num_vectors, "index should be less than %d, %d provided!\n",
-				num_vectors, i_col);
+			REQUIRE(i_row>=0, "Index %d negative!\n", i_row);
```
Comment (karlnapf) [inline src/shogun/lib/SGSparseMatrix.h]: Sorry I did not spot this earlier, but "Index" is ambiguous. Please use row or column index

**Comment by yingryic** [comment]
take a look @karlnapf

**Comment by karlnapf** [comment]
Thanks a lot! Really appreciated, this type of proper work

**Pull request merged**